bentoml/_internal/yatai_client/__init__.py

```diff
@@ -0,0 +1,263 @@
+                except Exception as e:
+                    finish_req = FinishUploadBentoSchema(
+                        status=BentoUploadStatus.FAILED,
+                        reason=str(e),
+                    )
+                yatai_rest_client.finish_upload_bento(bento_repository_name=bento_repository.name, version=version, req=finish_req)
+            if finish_req.status != BentoUploadStatus.SUCCESS:
+                logger.error(f'Upload {finish_req}: {finish_req.reason}')
+            else:
+                logger.info('Upload successfully')
+
+    @inject
+    def pull_bento(self, tag: t.Union[str, Tag], *, force: bool = False,
+                   bento_store: "BentoStore" = Provide[BentoMLContainer.bento_store],
+                   model_store: "ModelStore" = Provide[BentoMLContainer.model_store]) -> "Bento":
+        try:
+            bento = bento_store.get(tag)
+            if not force:
+                logger.info(f"Bento {tag} already exists locally, skipping pull")
+                return bento
+            bento_store.delete(tag)
+        except NotFound:
+            pass
+        _tag = Tag.from_taglike(tag)
+        yatai_rest_client = get_current_yatai_rest_api_client()
+        remote_bento = yatai_rest_client.get_bento(bento_repository_name=_tag.name, version=_tag.version)
+        with ThreadPoolExecutor(max_workers=max(len(remote_bento.manifest.models), 1)) as executor:
+            futures = executor.map(lambda x: self.pull_model(x, force=force), remote_bento.manifest.models)
+            list(futures)
+        remote_bento = yatai_rest_client.presign_bento_download_url(_tag.name, _tag.version)
+        url = remote_bento.presigned_download_url
+        response = requests.get(url, stream=True)
+        if response.status_code != 200:
+            raise BentoMLException(f"Failed to download bento {_tag}: {response.text}")
+        total_size_in_bytes = int(response.headers.get('content-length', 0))
+        block_size = 1024  # 1 Kibibyte
+        with NamedTemporaryFile() as tar_file:
+            with tqdm(desc=f'Downloading bento {_tag}', total=total_size_in_bytes, unit='iB', unit_scale=True) as progress_bar:
+                for data in response.iter_content(block_size):
+                    progress_bar.update(len(data))
+                    tar_file.write(data)
+            tar_file.seek(0, 0)
+            tar = tarfile.open(fileobj=tar_file, mode='r:gz')
+            with fs.open_fs('temp://') as temp_fs:
+                for member in tar.getmembers():
+                    f = tar.extractfile(member)
+                    if f is None:
+                        continue
+                    p = Path(member.name)
+                    if p.parent != Path('.'):
+                        temp_fs.makedirs(str(p.parent), recreate=True)
+                    temp_fs.writebytes(member.name, f.read())
+                Bento.save_models(temp_fs, remote_bento.manifest.models, model_store=model_store)
```

Comment:
I'm not so sure about this change here, the logic in `save_models` is pretty specific for Bento creation. I'd suggest undo the refactoring around `save_models` method, and change this line to:

```
bento = Bento.from_fs(temp_fs)
for model_tag in remote_bento.manifest.models:
    copy_model_if_not_found(model_tag, src=bento.model_store, target=model_store)
```
And for the utility function:
```
def copy_model_if_not_found(model_tag, src, target):
    """copy a model from src model store to taget modelstore, and do nothing if the model tag
    already exist in target model store
    """
    try:
        target_model_store.get(model_tag):  # if model tag already found in target
        return
    except NotFound:
         pass
 
    model = src_model_store.get(model_tag)
    model.save(model_store=target_model_store)

```